OpenSim/Common/Storage.cpp

```diff
@@ -42,20 +42,60 @@
 using namespace OpenSim;
 using namespace std;
 
-void convertTableToStorage(const TimeSeriesTable& table, Storage& sto) {
+void convertTableToStorage(const AbstractDataTable* table, Storage& sto)
+{
     sto.purge();
-    OpenSim::Array<std::string> labels("", (int)table.getNumColumns() + 1);
+    TimeSeriesTable out;
+
+    if (auto td = dynamic_cast<const TimeSeriesTable*>(table))
+        // Table is already flattened, so clone for further processing
+        out = TimeSeriesTable{ *td };
+    else if (auto tst = dynamic_cast<const TimeSeriesTable_<SimTK::Vec2>*>(table))
+        out = tst->flatten();
+    else if (auto tst = dynamic_cast<const TimeSeriesTable_<SimTK::Vec3>*>(table))
+        out = tst->flatten({ "_x", "_y", "_z" });
+    else if (auto tst = dynamic_cast<const TimeSeriesTable_<SimTK::Vec4>*>(table))
+        out = tst->flatten();
+    else if (auto tst = dynamic_cast<const TimeSeriesTable_<SimTK::Vec5>*>(table))
+        out = tst->flatten();
+    else if (auto tst = dynamic_cast<const TimeSeriesTable_<SimTK::Vec6>*>(table))
+        out = tst->flatten();
+    else if (auto tst = dynamic_cast<const TimeSeriesTable_<SimTK::Vec7>*>(table))
+        out = tst->flatten();
+    else if (auto tst = dynamic_cast<const TimeSeriesTable_<SimTK::Vec8>*>(table))
+        out = tst->flatten();
+    else if (auto tst = dynamic_cast<const TimeSeriesTable_<SimTK::Vec9>*>(table))
+        out = tst->flatten();
+    else if (auto tst = dynamic_cast<const TimeSeriesTable_<SimTK::Vec<10>>*>(table))
+        out = tst->flatten();
+    else if (auto tst = dynamic_cast<const TimeSeriesTable_<SimTK::Vec<11>>*>(table))
+        out = tst->flatten();
+    else if (auto tst = dynamic_cast<const TimeSeriesTable_<SimTK::Vec<12>>*>(table))
+        out = tst->flatten();
+    else if (auto tst = dynamic_cast<const TimeSeriesTable_<SimTK::UnitVec3>*>(table))
+        out = tst->flatten({ "_x", "_y", "_z" });
+    else if (auto tst = dynamic_cast<const TimeSeriesTable_<SimTK::Quaternion>*>(table))
+        out = tst->flatten();
```

Comment:
How would this scale if a new type of TimeSeriesTable_<x> is introduced?

Reply:
It doesn't scale automatically. We would have to add it here. Given how slow we have been to add new types (at the moment we do not even have Outputs that exercise the types that are handled) I don't think we will be needing to support new types soon . If we need more flexibility we may have to revisit, but I think it is unlikely. It is more likely that `Storage` will be deprecated by then.